usersession/agent/rest_api.go

```diff
@@ -46,3 +60,123 @@ func sessionInfo(c *Command, r *http.Request) Response {
 	}
 	return SyncResponse(m)
 }
+
+type serviceInstruction struct {
+	Action   string   `json:"action"`
+	Services []string `json:"services"`
+}
+
+var (
+	stopTimeout = time.Duration(timeout.DefaultTimeout)
+	killWait    = 5 * time.Second
+)
+
+func serviceStart(inst *serviceInstruction, sysd systemd.Systemd) Response {
+	// Refuse to start non-snap services
+	for _, service := range inst.Services {
+		if !strings.HasPrefix(service, "snap.") {
+			return InternalError("cannot start non-snap service %v", service)
+		}
+	}
+
+	startErrors := make(map[string]string)
+	var started []string
+	for _, service := range inst.Services {
+		if err := sysd.Start(service); err != nil {
+			startErrors[service] = err.Error()
+			break
+		}
+		started = append(started, service)
+	}
+	// If we got any failures, attempt to stop the services we started.
+	if len(startErrors) != 0 {
+		for _, service := range started {
+			if err := sysd.Stop(service, stopTimeout); err != nil {
+				startErrors[service] = err.Error()
```

Comment:
just recording these errors as start errors is a bit odd,  maybe we need 2 maps so that the client can decide how to treat them,  wrappers.StartServices treats them differently than base errors for examples

Reply:
I've separate out the errors into `start-errors` and `stop-errors` dictionaries, and updated the "stop" action to use use similar nesting in the error value for consistency.  This should also make it possible to add more information to the error later if we find a need.